src/gui_common/CustomDialog.cs

```diff
@@ -0,0 +1,489 @@
+﻿using System;
+using Godot;
+
+/// <summary>
+///   A reimplementation of WindowDialog for a much more customized style and functionality. Suitable for general use
+///   or as a base class for any custom window dialog derived types.
+/// </summary>
+/// <remarks>
+///   <para>
+///     This uses Tool attribute to make this class be run in the Godot editor for real-time feedback as this class
+///     handles UI visuals extensively through code. Not necessary but very helpful when editing scenes involving
+///     any custom dialogs.
+///   </para>
+/// </remarks>
+[Tool]
+public class CustomDialog : Popup, ICustomPopup
+{
+    private string windowTitle;
+
+    private bool closeHovered;
+
+    private Vector2 dragOffset;
+    private Vector2 dragOffsetFar;
+
+    private TextureButton closeButton;
+
+    private DragType dragType = DragType.None;
+
+    [Flags]
+    private enum DragType
+    {
+        None = 0,
+        Move = 1,
+        ResizeTop = 1 << 1,
+        ResizeRight = 1 << 2,
+        ResizeBottom = 1 << 3,
+        ResizeLeft = 1 << 4,
+    }
+
+    /// <summary>
+    ///   The text displayed in the window's title bar.
+    /// </summary>
+    [Export]
+    public string WindowTitle
+    {
+        get => windowTitle;
+        set
+        {
+            windowTitle = value;
+            MinimumSizeChanged();
+            Update();
+        }
+    }
+
+    /// <summary>
+    ///   If true, the user can resize the window.
+    /// </summary>
+    [Export]
+    public bool Resizable { get; set; }
+
+    /// <summary>
+    ///   If true, the window's position is clamped inside the screen so it doesn't go out of bounds.
+    /// </summary>
+    [Export]
+    public bool BoundToScreenArea { get; set; } = true;
+
+    [Export]
+    public bool ExclusiveAllowCloseOnEscape { get; private set; } = true;
```

Comment:
Shouldn't the set be public here as this is an export property (meant to be set from the outside)?